exp/internal/controllers/machinepool_controller.go

```diff
@@ -289,3 +294,80 @@ func (r *MachinePoolReconciler) reconcileDeleteExternal(ctx context.Context, m *
 	// Return true if there are no more external objects.
 	return len(objects) == 0, nil
 }
+
+func (r *MachinePoolReconciler) watchClusterNodes(ctx context.Context, cluster *clusterv1.Cluster) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	if !conditions.IsTrue(cluster, clusterv1.ControlPlaneInitializedCondition) {
+		log.V(5).Info("Skipping node watching setup because control plane is not initialized")
+		return nil
+	}
+
+	// If there is no tracker, don't watch remote nodes
+	if r.Tracker == nil {
+		return nil
+	}
+
+	return r.Tracker.Watch(ctx, remote.WatchInput{
+		Name:         "machinepool-watchNodes",
+		Cluster:      util.ObjectKey(cluster),
+		Watcher:      r.controller,
+		Kind:         &corev1.Node{},
+		EventHandler: handler.EnqueueRequestsFromMapFunc(r.nodeToMachinePool),
+	})
+}
+
+func (r *MachinePoolReconciler) nodeToMachinePool(o client.Object) []reconcile.Request {
+	node, ok := o.(*corev1.Node)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Node but got a %T", o))
+	}
+
+	var filters []client.ListOption
+	// Match by clusterName when the node has the annotation.
+	if clusterName, ok := node.GetAnnotations()[clusterv1.ClusterNameAnnotation]; ok {
+		filters = append(filters, client.MatchingLabels{
+			clusterv1.ClusterNameLabel: clusterName,
+		})
+	}
+
+	// Match by namespace when the node has the annotation.
+	if namespace, ok := node.GetAnnotations()[clusterv1.ClusterNamespaceAnnotation]; ok {
+		filters = append(filters, client.InNamespace(namespace))
+	}
+
+	// Match by nodeName and status.nodeRef.name.
+	machinePoolList := &expv1.MachinePoolList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machinePoolList,
+		append(filters, client.MatchingFields{index.MachinePoolNodeNameField: node.Name})...); err != nil {
```

Comment:
I'm not entirely sure how to get indexing to work for MachinePools given there are multiple nodes per MachinePool... The extract func [here](https://github.com/kubernetes-sigs/cluster-api/pull/8443/files#diff-529ecbd30234264d79e067514cf018da592ab083d2045f4ff6828db54b9f2779R54) returns a list of all node names associated with the MachinePool. What should this filter be?

Reply:
From the documentation looks like this way using the filter should work as `MatchingFields` also [supports cache indices](https://github.com/ykakarap/controller-runtime/blob/60af59f5b22335516850ca11c974c8f614d5d073/pkg/client/options.go#L564).

And, indexing with multiple "keys" is apparently supported however it would [lack compatibility with Kubernetes API server](https://github.com/ykakarap/controller-runtime/blob/60af59f5b22335516850ca11c974c8f614d5d073/pkg/client/interfaces.go#L196-L198) (not sure what the means or what the impact would be).